libass/ass_threading.h

```diff
@@ -0,0 +1,360 @@
+
+#ifndef atomic_compare_exchange_weak_explicit
+#define atomic_compare_exchange_weak_explicit atomic_compare_exchange_strong_explicit
+#endif
+
+#ifndef atomic_store_explicit
+#define atomic_store_explicit(obj, desired, order) ((*(obj) = (desired)), MAYBE_FENCE(order))
+#endif
+
+#ifndef atomic_load_explicit
+#define atomic_load_explicit(obj, order) (MAYBE_FENCE(order), *(obj))
+#endif
+
+#ifndef atomic_fetch_add_explicit
+#define atomic_fetch_add_explicit(obj, arg, order) INTERLOCKED(ExchangeAdd, order)(obj, arg)
+#endif
+
+#ifndef atomic_fetch_sub_explicit
+#define atomic_fetch_sub_explicit(obj, arg, order) INTERLOCKED(ExchangeAdd, order)(obj, -(AtomicInt)arg)
+#endif
+
+#ifndef atomic_init
+#define atomic_init(obj, arg) (*(obj) = (arg))
+#endif
+
+#ifndef atomic_exchange_explicit
+#define atomic_exchange_explicit(obj, desired, order) INTERLOCKED(Exchange, order)(obj, desired)
+#endif
+
+#ifndef _Atomic
+#define _Atomic
+#endif
+
+#else // HAVE_STDATOMIC_H/_WIN32
+
+#define HAVE_ATOMICS 0
+
+typedef intptr_t AtomicInt;
+
+#ifndef atomic_compare_exchange_strong_explicit
+#define atomic_compare_exchange_strong_explicit(obj, expected, desired, order, order2) \
+    ((*(obj) == *(expected)) ? \
+        (*(obj) = (desired), true) : \
+        (*(expected) = *(obj), false))
+#endif
+
+#ifndef atomic_compare_exchange_weak
+#define atomic_compare_exchange_weak_explicit atomic_compare_exchange_strong_explicit
+#endif
+
+#ifndef atomic_store_explicit
+#define atomic_store_explicit(obj, desired, order) (*(obj) = (desired))
+#endif
+
+#ifndef atomic_load_explicit
+#define atomic_load_explicit(obj, order) (*(obj))
+#endif
+
+#define FETCH_OP(name, op) \
+static inline intptr_t do_fetch_ ## name(intptr_t *obj, intptr_t arg) \
+{ \
+    intptr_t ret = *obj; \
+    *obj = *obj op arg; \
+    return ret; \
+}
+
+#ifndef atomic_fetch_add_explicit
+FETCH_OP(add, +)
+#define atomic_fetch_add_explicit(obj, arg, order) do_fetch_add(obj, arg)
+#endif
+
+#ifndef atomic_fetch_sub_explicit
+FETCH_OP(sub, -)
+#define atomic_fetch_sub_explicit(obj, arg, order) do_fetch_sub(obj, arg)
+#endif
+
+#ifndef atomic_init
+#define atomic_init(obj, arg) (*(obj) = (arg))
+#endif
+
+#ifndef atomic_exchange_explicit
+static inline intptr_t do_exchange(intptr_t *obj, intptr_t desired) {
+    intptr_t old = *obj;
+    *obj = desired;
+    return old;
+}
+
+#define atomic_exchange_explicit(obj, desired, order) do_exchange(obj, desired)
+#endif
+
+#ifndef _Atomic
+#define _Atomic
+#endif
+
+#endif // HAVE_STDATOMIC_H/_WIN32
+
+#if HAVE_ATOMICS
+
+#ifdef CONFIG_W32THREAD
+
+#define ENABLE_THREADS 1
+
+struct ThreadStruct {
+    void *(*start_routine)(void *);
+    void *arg;
+    void *ret;
+    HANDLE handle;
+};
+
+typedef struct ThreadStruct *pthread_t;
+
+typedef CRITICAL_SECTION   pthread_mutex_t;
+typedef CONDITION_VARIABLE pthread_cond_t;
+
+static inline int pthread_mutex_init(pthread_mutex_t *mtx, const void *attr)
+{
+    assert(!attr);
+    InitializeCriticalSection(mtx);
```

Comment:
If recursive mutexes are not needed, consider using SWR locks instead. They are lighter and faster.

Reply:
Interesting; it looks like SRW locks perform better in *most* (but not all) cases in this benchmark, but the results appear fairly noisy, and are highly dependent on exactly how much contention is happening. I'd still definitely be curious to see a measured impact in a real-world libass scenario.